Update Checkpoints Feature (Expand tooltip and icons, add slash commands, refactoring)

Wolfsblvt opened the pull request:
Improved the existing checkpoints feature by a lot. (I hope)
Now that I figured out what checkpoints actually meant and how they were supposed to be used, I liked them a lot. So I figured this could be useful.

- Remove double flag icon on messages
  If a checkpoint exists, the new flag will be present even if message buttons are hidden (like before). Overwriting checkpoint functionality is moved to Shift+Click.
- Add checkpoint name to the tooltip of the checkpoint flag
- Added slash commands for checkpoints and branches.
  I likely made several typos. Feel free to correct them.
  - `/branch-create` - Creates a branch for a specified message
  - `/checkpoint-create` - Creates a checkpoint for a specified message
  - `/checkpoint-go` - Opens the checkpoint chat behind a specified message
  - `/checkpoint-exit` - Exits out of a checkpoint chat, returning to the parent
  - `/checkpoint-parent` - Returning the name of the parent chat, if current chat is a checkpoint
  - `/checkpoint-get` - Gets the name of a checkpoint tied to a specified message
  - `/checkpoint-list` - Lists all available checkpoints inside this chat
- Fixed toast saying referring to last message, instead of any message
- Removed some legacy code, refactored checkpoint popups to new popups, some more code cleanup
- Moved all checkpoints/branch code to the `bookmarks.js` file

## Checklist:

- [x] I have read the [Contributing guidelines](https://www.youtube.com/watch?v=dQw4w9WgXcQ).

Wolfsblvt committed:
Update toeast message

Wolfsblvt committed:
Remove double flags for bookmarks

- Remove the double flag if a checkpoint exists
- Add the "replace checkpoint" functionality as a Shift+Click option to the checkpoint flag

Wolfsblvt committed:
Remove legacy bookmark messages

Wolfsblvt committed:
Refactor checkpoint popups

Wolfsblvt committed:
Expand checkpoint tooltip to contain name

Wolfsblvt committed:
Refactor move bookmark functionality

Wolfsblvt committed:
Refactor bookmarks with correct async and export

Wolfsblvt committed:
Add slash commands for checkpoints and branch

Cohee1207 committed:
Don't use classic Mac line endings

Cohee1207 committed:
Unbreak timelines

Cohee1207 commented:
Checkpoints name auto-generation is broken. Leaving the name empty aborts checkpoint creation.

Cohee1207 commented on public/scripts/bookmarks.js:
```diff
@@ -351,8 +370,314 @@ async function convertSoloToGroupChat() {
+        /** @param {{links?: string}} args @returns {Promise<string>} */
+        callback: async (args, _) => {
+            const result = [];
+            for (const mesId in chat) {
```
This feels wrong. Just use a classic for i < length loop.

Wolfsblvt replied:
I blame copilot. Will fix later tonight. Makes sense.

Wolfsblvt replied:
Easier and cleaner with a filter+map anyway.

Cohee1207 commented on public/scripts/bookmarks.js:
```diff
@@ -351,8 +370,314 @@ async function convertSoloToGroupChat() {
     $('#option_convert_to_group').on('click', convertSoloToGroupChat);
-});
+
+    $(document).on('click', '.select_chat_block, .bookmark_link, .mes_bookmark', async function (e) {
```
`.bookmark_link` class is dead and could only be encountered in very (VERY) old chats. Doubt anyone still cares legacy system messages for checkpoint navigation.

Cohee1207 commented on public/scripts/bookmarks.js:
```diff
@@ -351,8 +370,314 @@ async function convertSoloToGroupChat() {
+            const mainChat = await backToMainChat();
+            return mainChat ?? '';
+        },
+        helpString: 'Exit the checkpoint chat.<br />If not in a checkpoint chat, returns null.',
```
It does not, in fact, return null.

Wolfsblvt replied:
Fixed that and the few other places. Please check for any other wording issues.

Cohee1207 commented on public/scripts/bookmarks.js:
```diff
@@ -57,13 +56,11 @@ async function getExistingChatNames() {
-    let name = await callPopup(popupText, 'input');
 
-    if (name === false) {
+    let name = forceName || await Popup.show.input('Create Checkpoint', '<span class="margin-right-10px">Enter Checkpoint Name:</span><small>(Leave empty to auto-generate)</small>');
```
Move the popup body to a named HTML template.

Wolfsblvt replied:
Done and expanded

Cohee1207 commented on public/scripts/bookmarks.js:
```diff
@@ -215,10 +216,24 @@ async function createNewBookmark(mesId) {
+ */
+export function updateBookmarkDisplay(mes, newBookmarkLink = null) {
+    newBookmarkLink && mes.attr('bookmark_link', newBookmarkLink);
+    mes.find('.mes_bookmark').attr('title', `Checkpoint\n${mes.attr('bookmark_link')}\n\n${mes.find('.mes_bookmark').data('tooltip')}`);
```
Save `.mes_bookmark` to a variable not to query for it twice.

Wolfsblvt replied:
I did. Then I thought it's totally not needed for such a small thing and reverted. But yeah...

Cohee1207 replied:
Message rendering is a bit of a hotspot, so it would make sense to optimize it.

Cohee1207 commented on public/scripts/bookmarks.js:
```diff
@@ -351,8 +370,314 @@ async function convertSoloToGroupChat() {
+        name: 'branch-create',
+        returns: 'Name of the new branch',
+        callback: async (args, text) => {
+            const mesId = Number(args.mesId ?? text ?? getLastMessageId());
```
Message Id is both a named and unnamed argument? This is confusing. Pick one.
Also `text` would never be null/undefined.

Wolfsblvt replied:
I was looking at quite a few other commands, where we supported both unnamed arguments or named arguments for the same value. `/getvar` might be the most prominent example.
Though for a new slash command, there is no need to follow that principle, that is true...

If I make it unnamed, it's easier to type, sure. If I only make it named, we have a command with no unnamed arguments. Weird. But it's consistent, because the other commands here all take a `mes` argument for the ID, instead of unnamed argument.
Meeeeh.

Cohee1207 replied:
If had a choice, I'd say - leave the unnamed, drop the named.

Wolfsblvt replied:
Removed all named if message is the only argument that can be provided.

Cohee1207 commented on public/scripts/bookmarks.js:
```diff
@@ -351,8 +370,314 @@ async function convertSoloToGroupChat() {
+
+            return checkPointName;
+        },
+        namedArgumentList: [
```
Same here. 𝓒𝓸𝓷𝓼𝓲𝓼𝓽𝓮𝓷𝓬𝔂

Wolfsblvt committed:
Merge branch 'staging' into small-bookmark-updates

Wolfsblvt committed:
Changes based on code review

Wolfsblvt committed:
Remove unnecessary double popup for replace

Wolfsblvt commented:
Fixed all mentions, and did a small change, removing the "Replace" popup. With the shift-click already being an explicit action, no need to confirm again.

Ready for testing.

Cohee1207 committed:
Merge branch 'staging' into small-bookmark-updates

Cohee1207 reviewed (changes requested):
> Checkpoints name auto-generation is broken. Leaving the name empty aborts checkpoint creation.

This is still a thing.

STR:
1. Press checkpoint flag
2. Leave the field empty
3. Click "save"

ER: checkpoint created
AR: no checkpoint created

Wolfsblvt commented:
Dear god. I fixed like a dozen things, I totally missed this one. Should not be difficult, lol.
On it.

Wolfsblvt committed:
Fix checkpoint create auto generate (and more)

- Fix empty checkpoint name on mes click to auto generate
- Change /checkpoint-create to not require name and auto generate
- Fix popup input type returning empty string instead of null on empty input field

Wolfsblvt commented:
I expect this to take another two weeks until this is ready to merge 🙃

Cohee1207 committed:
Update /caption mesId argument name

Wolfsblvt committed:
Fix /checkpoint-create "mesId" argument

Cohee1207 committed:
Update old code style

Wolfsblvt committed:
Fix /checkpoints-list with ids

Cohee1207 committed:
Merge branch 'staging' into small-bookmark-updates

Cohee1207 committed:
Fix shift+click in chat manager

Cohee1207 reviewed (approved):
LGTM. I couldn't find any more problems.

Cohee1207 committed:
Fix translation key removal

Cohee1207 committed:
Fix TTS falsy checks
